Microprofile Metric Missing "Scope" Tag Only When Using Constructor Injection

### Describe the bug

When injecting multiple microprofile metrics into a Quarkus managed bean using constructor injection, only the first such metric has the expected "scope=application" tag. Simply changing the bean to use field injection instead of constructor injection works around this bug.

### Expected behavior

Microprofile metrics spec states that metrics should include a "scope" tag.

### Actual behavior

The first microprofile counter injected into the constructor has the "scope" tag, but subsequent metrics injected this way lack the "scope" tag.

### How to Reproduce?

[microprofile-metric-tags-ctor-injection.tar.gz](https://github.com/quarkusio/quarkus/files/13604621/microprofile-metric-tags-ctor-injection.tar.gz)

The attached project constructor injects two microprofile counters.

```java
@Path("/hello")
public class GreetingResource {

  private final Counter first;
  private final Counter second;

  @Inject
  public GreetingResource(
      @Metric(name = "first-counter") final Counter first,
      @Metric(name = "second-counter") final Counter second) {
    this.first = Objects.requireNonNull(first);
    this.second = Objects.requireNonNull(second);
  }
```

the QuarkusTest included in the project fails, because the second counter lacks the "scope" tag.

```
java.lang.AssertionError: Expected to find a tag with key 'scope' on counter MeterId{name='second-counter', tags=[]}

	at org.acme.GreetingResourceTest.lambda$counter_has_expected_tags$2(GreetingResourceTest.java:30)
	at java.base/java.util.Optional.ifPresentOrElse(Optional.java:198)
	at org.acme.GreetingResourceTest.counter_has_expected_tags(GreetingResourceTest.java:25)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
```

### Output of `uname -a` or `ver`

Darwin pixee-mbp-gilday.localdomain 23.1.0 Darwin Kernel Version 23.1.0: Mon Oct  9 21:27:24 PDT 2023; root:xnu-10002.41.9~6/RELEASE_ARM64_T6000 arm64

### Output of `java -version`

openjdk version "17.0.9" 2023-10-17 OpenJDK Runtime Environment Temurin-17.0.9+9 (build 17.0.9+9) OpenJDK 64-Bit Server VM Temurin-17.0.9+9 (build 17.0.9+9, mixed mode)

### Quarkus version or git rev

3.6.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.6 (bc0240f3c744dd6b6ec2920b3cd08dcc295161ae) Maven home: /Users/jgilday/.m2/wrapper/dists/apache-maven-3.9.6-bin/3311e1d4/apache-maven-3.9.6 Java version: 17.0.9, vendor: Eclipse Adoptium, runtime: /Library/Java/JavaVirtualMachines/temurin-17.jdk/Contents/Home Default locale: en_US, platform encoding: UTF-8 OS name: "mac os x", version: "14.1.2", arch: "aarch64", family: "mac"

### Additional information

_No response_

--- a/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/binder/mpmetrics/MetricRegistryAdapter.java
+++ b/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/binder/mpmetrics/MetricRegistryAdapter.java
@@ -90,13 +90,13 @@ public class MetricRegistryAdapter implements MetricRegistry {
 
     Counter interceptorCounter(Metadata metadata, String... tags) {
         return internalCounter(internalGetMetadata(metadata, MetricType.COUNTER),
-                new MetricDescriptor(metadata.getName(), tags));
+                new MetricDescriptor(metadata.getName(), scopeTags(tags)));
     }
 
     Counter injectedCounter(org.eclipse.microprofile.metrics.annotation.Metric annotation) {
         return internalCounter(
                 internalGetMetadata(annotation.name(), MetricType.COUNTER).merge(annotation),
-                new MetricDescriptor(annotation.name(), annotation.tags()));
+                new MetricDescriptor(annotation.name(), scopeTags(annotation.tags())));
     }
 
     CounterAdapter internalCounter(MpMetadata metadata, MetricDescriptor id) {
@@ -138,13 +138,13 @@ public class MetricRegistryAdapter implements MetricRegistry {
 
     ConcurrentGaugeImpl interceptorConcurrentGauge(Metadata metadata, String... tags) {
         return internalConcurrentGauge(internalGetMetadata(metadata, MetricType.CONCURRENT_GAUGE),
-                new MetricDescriptor(metadata.getName(), tags));
+                new MetricDescriptor(metadata.getName(), scopeTags(tags)));
     }
 
     ConcurrentGaugeImpl injectedConcurrentGauge(org.eclipse.microprofile.metrics.annotation.Metric annotation) {
         return internalConcurrentGauge(
                 internalGetMetadata(annotation.name(), MetricType.CONCURRENT_GAUGE).merge(annotation),
-                new MetricDescriptor(annotation.name(), annotation.tags()));
+                new MetricDescriptor(annotation.name(), scopeTags(annotation.tags())));
     }
 
     ConcurrentGaugeImpl internalConcurrentGauge(MpMetadata metadata, MetricDescriptor id) {
@@ -276,7 +276,7 @@ public class MetricRegistryAdapter implements MetricRegistry {
     HistogramAdapter injectedHistogram(org.eclipse.microprofile.metrics.annotation.Metric annotation) {
         return internalHistogram(
                 internalGetMetadata(annotation.name(), MetricType.HISTOGRAM).merge(annotation),
-                new MetricDescriptor(annotation.name(), annotation.tags()));
+                new MetricDescriptor(annotation.name(), scopeTags(annotation.tags())));
     }
 
     HistogramAdapter internalHistogram(MpMetadata metadata, MetricDescriptor id) {
@@ -319,7 +319,7 @@ public class MetricRegistryAdapter implements MetricRegistry {
     MeterAdapter injectedMeter(org.eclipse.microprofile.metrics.annotation.Metric annotation) {
         return internalMeter(
                 internalGetMetadata(annotation.name(), MetricType.METERED).merge(annotation),
-                new MetricDescriptor(annotation.name(), annotation.tags()));
+                new MetricDescriptor(annotation.name(), scopeTags(annotation.tags())));
     }
 
     MeterAdapter internalMeter(MpMetadata metadata, MetricDescriptor id) {
@@ -363,12 +363,12 @@ public class MetricRegistryAdapter implements MetricRegistry {
     TimerAdapter injectedTimer(org.eclipse.microprofile.metrics.annotation.Metric annotation) {
         return internalTimer(
                 internalGetMetadata(annotation.name(), MetricType.TIMER).merge(annotation),
-                new MetricDescriptor(annotation.name(), annotation.tags()));
+                new MetricDescriptor(annotation.name(), scopeTags(annotation.tags())));
     }
 
     TimerAdapter interceptorTimer(Metadata metadata, String... tags) {
         return internalTimer(internalGetMetadata(metadata, MetricType.TIMER),
-                new MetricDescriptor(metadata.getName(), tags));
+                new MetricDescriptor(metadata.getName(), scopeTags(tags)));
     }
 
     TimerAdapter internalTimer(MpMetadata metadata, MetricDescriptor id) {
@@ -465,7 +465,7 @@ public class MetricRegistryAdapter implements MetricRegistry {
     TimerAdapter injectedSimpleTimer(org.eclipse.microprofile.metrics.annotation.Metric annotation) {
         return internalSimpleTimer(
                 internalGetMetadata(annotation.name(), MetricType.SIMPLE_TIMER).merge(annotation),
-                new MetricDescriptor(annotation.name(), annotation.tags()));
+                new MetricDescriptor(annotation.name(), scopeTags(annotation.tags())));
     }
 
     TimerAdapter internalSimpleTimer(MpMetadata metadata, MetricDescriptor id) {
@@ -657,12 +657,21 @@ public class MetricRegistryAdapter implements MetricRegistry {
         return null;
     }
 
+    Tags scopeTags() {
+        return Tags.of("scope", this.type.getName());
+    }
+
     Tags scopeTags(Tag... tags) {
-        Tags out = Tags.of("scope", this.type.getName());
+        Tags out = scopeTags();
         for (Tag t : tags) {
             out = out.and(t.getTagName(), t.getTagValue());
         }
         return out;
+    }
+
+    Tags scopeTags(String... tags) {
+        Tags in = Tags.of(tags);
+        return scopeTags().and(in);
     }
 
     private MpMetadata internalGetMetadata(String name, MetricType type) {

--- a/integration-tests/micrometer-mp-metrics/src/main/java/io/quarkus/it/micrometer/mpmetrics/MessageResource.java
+++ b/integration-tests/micrometer-mp-metrics/src/main/java/io/quarkus/it/micrometer/mpmetrics/MessageResource.java
@@ -1,34 +1,60 @@
 package io.quarkus.it.micrometer.mpmetrics;
 
+import java.util.Collection;
+import java.util.Objects;
+
 import jakarta.ws.rs.GET;
 import jakarta.ws.rs.Path;
 import jakarta.ws.rs.PathParam;
 
+import org.eclipse.microprofile.metrics.Counter;
+import org.eclipse.microprofile.metrics.annotation.Metric;
+
+import io.micrometer.core.instrument.Meter;
 import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.search.Search;
 
 @Path("/message")
 public class MessageResource {
 
     private final MeterRegistry registry;
+    private final Counter first;
+    private final Counter second;
 
-    public MessageResource(MeterRegistry registry) {
+    public MessageResource(MeterRegistry registry,
+            @Metric(name = "first-counter") final Counter first,
+            @Metric(name = "second-counter") final Counter second) {
         this.registry = registry;
+        this.first = Objects.requireNonNull(first);
+        this.second = Objects.requireNonNull(second);
     }
 
     @GET
     public String message() {
+        first.inc();
+        second.inc();
         return registry.getClass().getName();
     }
 
     @GET
     @Path("fail")
     public String fail() {
+        first.inc();
         throw new NullPointerException("Failed on purpose");
     }
 
     @GET
     @Path("item/{id}")
     public String item(@PathParam("id") String id) {
+        second.inc();
         return "return message with id " + id;
+    }
+
+    @GET
+    @Path("mpmetrics")
+    public String metrics() {
+        Collection<Meter> meters = Search.in(registry).name(s -> s.contains("mpmetrics")).meters();
+        meters.addAll(Search.in(registry).name(s -> s.endsWith("-counter")).meters());
+        return meters.stream().allMatch(x -> x.getId().getTag("scope") != null) ? "OK" : "FAIL";
     }
 }

--- a/integration-tests/micrometer-mp-metrics/src/test/java/io/quarkus/it/micrometer/mpmetrics/MPMetricsTest.java
+++ b/integration-tests/micrometer-mp-metrics/src/test/java/io/quarkus/it/micrometer/mpmetrics/MPMetricsTest.java
@@ -70,7 +70,10 @@ class MPMetricsTest {
                         "io_quarkus_it_micrometer_mpmetrics_PrimeResource_highestPrimeNumberSoFar2{scope=\"application\"} 887.0"))
 
                 // the counter associated with a timed method should have been removed
-                .body(not(containsString("io_quarkus_it_micrometer_mpmetrics_PrimeResource_checkPrime")));
+                .body(not(containsString("io_quarkus_it_micrometer_mpmetrics_PrimeResource_checkPrime")))
+
+                // no calls to /message
+                .body(not(containsString("/message")));
     }
 
     @Test
@@ -83,10 +86,28 @@ class MPMetricsTest {
     }
 
     @Test
-    @Order(8)
+    @Order(6)
     void callMessage() {
         given()
                 .when().get("/message")
+                .then()
+                .statusCode(200);
+    }
+
+    @Test
+    @Order(7)
+    void callMessageFail() {
+        given()
+                .when().get("/message/fail")
+                .then()
+                .statusCode(500);
+    }
+
+    @Test
+    @Order(8)
+    void callMessageId() {
+        given()
+                .when().get("/message/item/35")
                 .then()
                 .statusCode(200);
     }
@@ -106,7 +127,10 @@ class MPMetricsTest {
                         "highestPrimeNumberSoFar 887.0"))
                 .body(containsString(
                         "io_quarkus_it_micrometer_mpmetrics_InjectedInstance_notPrime_total{scope=\"application\"}"))
-                .body(not(containsString("/message")));
+                .body(containsString(
+                        "first_counter_total{scope=\"application\"}"))
+                .body(containsString(
+                        "second_counter_total{scope=\"application\"}"));
     }
 
     @Test
@@ -123,4 +147,13 @@ class MPMetricsTest {
                         Matchers.equalTo(887.0f));
     }
 
+    @Test
+    @Order(11)
+    void meters() {
+        given()
+                .when().get("/message/mpmetrics")
+                .then()
+                .statusCode(200)
+                .log().body();
+    }
 }

